vsp.go

```diff
@@ -57,65 +61,309 @@ func (mw *MultiWallet) NewVSPClient(vspHost string, walletID int, purchaseAccoun
+
+	go func() {
+		log.Infof("Running ticket buyer on Wallet: %s", vsp.w.Name)
+
+		ctx, cancel := vsp.w.shutdownContextWithCancel()
+		vsp.w.cancelAutoTicketBuyer = cancel
+
+		err := vsp.runTicketBuyer(ctx, balanceToMaintain, passphrase)
+		if err != nil {
+			log.Errorf("Ticket buyer instance errored: %v", err)
+		}
+
+		vsp.w.cancelAutoTicketBuyer = nil
+	}()
+
+	return nil
+}
+
+func (vsp *VSP) runTicketBuyer(ctx context.Context, balanceToMaintain int64, passphrase []byte) error {
+	c := vsp.w.Internal().NtfnServer.MainTipChangedNotifications()
+	defer c.Done()
+
+	ctx, outerCancel := context.WithCancel(ctx)
+	defer outerCancel()
+	var fatal error
+	var fatalMu sync.Mutex
+
+	var nextIntervalStart, expiry int32
+	var cancels []func()
+	for {
+		select {
+		case <-ctx.Done():
+			defer outerCancel()
+			fatalMu.Lock()
+			err := fatal
+			fatalMu.Unlock()
+			if err != nil {
+				return err
+			}
+			return ctx.Err()
+		case n := <-c.C:
+			if len(n.AttachedBlocks) == 0 {
+				continue
+			}
+
+			tip := n.AttachedBlocks[len(n.AttachedBlocks)-1]
+			w := vsp.w.Internal()
+
+			// Don't perform any actions while transactions are not synced through
+			// the tip block.
+			rp, err := w.RescanPoint(ctx)
+			if err != nil {
+				return err
+			}
+			if rp != nil {
+				log.Debugf("Skipping autobuyer actions: transactions are not synced on Wallet: %s", vsp.w.Name)
+				continue
+			}
+
+			tipHeader, err := w.BlockHeader(ctx, tip)
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+			height := int32(tipHeader.Height)
+
+			// Cancel any ongoing ticket purchases which are buying
+			// at an old ticket price or are no longer able to
+			// create mined tickets the window.
+			if height+2 >= nextIntervalStart {
+				for i, cancel := range cancels {
+					cancel()
+					cancels[i] = nil
+				}
+				cancels = cancels[:0]
+
+				intervalSize := int32(w.ChainParams().StakeDiffWindowSize)
+				currentInterval := height / intervalSize
+				nextIntervalStart = (currentInterval + 1) * intervalSize
+
+				// Skip this purchase when no more tickets may be purchased in the interval and
+				// the next sdiff is unknown.  The earliest any ticket may be mined is two
+				// blocks from now, with the next block containing the split transaction
+				// that the ticket purchase spends.
+				if height+2 == nextIntervalStart {
+					log.Debugf("Skipping purchase: next sdiff interval starts soon")
+					continue
+				}
+				// Set expiry to prevent tickets from being mined in the next
+				// sdiff interval.  When the next block begins the new interval,
+				// the ticket is being purchased for the next interval; therefore
+				// increment expiry by a full sdiff window size to prevent it
+				// being mined in the interval after the next.
+				expiry = nextIntervalStart
+				if height+1 == nextIntervalStart {
+					expiry += intervalSize
+				}
+			}
+
+			cancelCtx, cancel := context.WithCancel(ctx)
+			cancels = append(cancels, cancel)
+			go func() {
```

Comment:
Based on recent conversation regarding this feature, this will need to more closely resemble dcrwallet's pattern here:
https://github.com/decred/dcrwallet/blob/9a8688e60b08b68b5952bb717122669cbb3e0989/ticketbuyer/tb.go#L199-L201

That should look like this in our case:
```
			// multiple is the max amount of tickets that can be bought based on available funds
			// calculate it here and start separate purchase goroutines for as many tickets that can be purchased
			// each goroutine MUST only buy 1 ticket.
			for i := 0; i < multiple; i++ {
				go buyTickets()
			}
```